codepropertygraph/src/main/scala/io/shiftleft/utils/TimeMetric.scala

```diff
@@ -0,0 +1,349 @@
+    if (alsoLogConsole && (!subStage || !supressSubstages))
+      println(msg)
+    baseLogger.debug(msg)
+  }
+  private val baseLogger: Logger             = LoggerFactory.getLogger(getClass)
+  private var recorder: Option[TimeRecorder] = None
+  private var fullName: Boolean              = false
+  private var maxBuffers: MaxPrintBuffers    = MaxPrintBuffers()
+  private var supressSubstages: Boolean      = true
+  private lazy val maxStartingBuffer = {
+    if (fullName) {
+      maxBuffers.MAX_FULLNAME_BUFFER + maxBuffers.MAX_DONE_TIME_BUFFER
+    } else {
+      maxBuffers.MAX_NAME_BUFFER + maxBuffers.MAX_DONE_TIME_BUFFER
+    }
+  }
+  private lazy val maxEndingBuffer = {
+    if (fullName) {
+      maxBuffers.MAX_FULLNAME_BUFFER
+    } else {
+      maxBuffers.MAX_NAME_BUFFER
+    }
+  }
+  private val osBean = ManagementFactory.getOperatingSystemMXBean().asInstanceOf[OperatingSystemMXBean]
+
+  def initialize(
+    useFullName: Boolean = false,
+    consoleLog: Boolean = false,
+    timeMetricRecordConfig: Option[TimeMetricRecordConfig] = None,
+    maxPrintBuffers: MaxPrintBuffers = MaxPrintBuffers(),
+    supressSubStagesConsoleLog: Boolean = true
+  ): Unit = {
+    fullName = useFullName
+    maxBuffers = maxPrintBuffers
+    supressSubstages = supressSubStagesConsoleLog
+    alsoLogConsole = consoleLog
+    timeMetricRecordConfig match {
+      case Some(config) =>
+        stagePerformance += (ACROSS_PROCESS_STAGE -> NumberProcessor())
+        recorder = Some(TimeRecorder(config))
+        recorder.get.start()
+      case _ =>
+    }
+  }
+
+  def setSupressSubstagesFlag(flag: Boolean): Unit = supressSubstages = flag
+
+  private class TimeRecorder(timeMetricRecordConfig: TimeMetricRecordConfig) extends Thread {
+
+    val queue: util.Queue[String] = new ConcurrentLinkedQueue[String]()
+
+    def pushEvent(event: String): Unit = queue.offer(event)
+
+    override def run(): Unit = {
+      val writer = new PrintWriter(timeMetricRecordConfig.resultFile)
+      try {
+        writer.println(s"DateTime, MainStage, SubStage, MetaData, CPU (%), Used Memory (MB), Max Heap (MB)")
+        while (true) {
+          // This while loop will take care of writing all the accumulated events in the queue first.
+          while (!queue.isEmpty) writer.println(queue.poll)
+          val (usedMemory, maxHeap, cpuUsed, _) = getCurrentCpuMemoryStats()
+          stagePerformance(ACROSS_PROCESS_STAGE).process(cpuUsed, usedMemory)
+          if (lastStageName != STAGE_NOT_SET) then stagePerformance (s"+ $lastStageName").process(cpuUsed, usedMemory)
+          if (lastSubStageName != STAGE_NOT_SET)
+            then stagePerformance (s"-- $lastSubStageName").process(cpuUsed, usedMemory)
+          writer.println(s"${getNewTime()}, $lastStageName, $lastSubStageName, $additionalMetaData, ${String
+              .format("%.2f", cpuUsed)}, ${String.format("%.2f", usedMemory)}, ${String.format("%.2f", maxHeap)}")
```

Comment:
Just wondering about locales, this may make a decimal `0,2` instead of `0.2` for example, which may break the CSV. I've encountered this often with my system local setting. This can be worked around by doing `"%.2f".formatLocal(java.util.Locale.US, usedMemory) `, which forces using a period instead of a comma.

Alternatively, you could wrap every column value in quotes

Reply:
It may make sense to set up a helper method for formatting the double strings, since they are done all over the code

Reply:
beyond 2 decimals values just become noise in the report. In fact, we can even exclude the decimals and only show the numbers as that makes more sense.

Reply:
done